Foundational work to get a cluster of pools for integration testing

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

N/A

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->
- 5a263bc2f6db213ca9a222968525607b0a45e117
  :round_pushpin: **move content of 'Test.Utils.Ports' to 'Cardano.Wallet.Network.Ports'**
  
- 6ed9efca634aa960137d1966be49215be7bede0f
  :round_pushpin: **rework the shelley launcher code so that we can start a cluster of nodes**
  using 'withCluster ... ... 0' will result in the same behavior as before. Right now, pools would
exists as process but not on chain. In order to exist on chain, pools need to be registered, which
will be done in a separate commit.

- 1a8b720a95870a0834e737f64656fd0ce597dfbc
  :round_pushpin: **fix database path**
  
- 6745ea944d20752353d6e4c978b6670ae7b75be2
  :round_pushpin: **Regenerate nix**
  
- 6b7b94763e4d3e4ceb9663dee0beb530b0a49e15
  :round_pushpin: **assign sensible default ordering to 'listAllTransactions'**
  
# Comments

<!-- Additional comments or screenshots to attach if any -->

This should make it possible to run a cluster of pools as simply as:

```hs
        let numberOfPools = 3
        withCluster tr Info numberOfPools $ \socketPath block0 (gp,vData) -> ...
```

So far, the pools won't have any _stake_ and won't therefore be any useful.. I'll give them some stake in another PR. 

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
move content of 'Test.Utils.Ports' to 'Cardano.Wallet.Network.Ports'

KtorZ committed:
rework the shelley launcher code so that we can start a cluster of nodes

using 'withCluster ... ... 0' will result in the same behavior as before. Right now, pools would
exists as process but not on chain. In order to exist on chain, pools need to be registered, which
will be done in a separate commit.

KtorZ committed:
fix database path

Commit:
Regenerate nix

KtorZ committed:
assign sensible default ordering to 'listAllTransactions'

paweljakubas commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -163,152 +179,250 @@ parseGenesisData = \case
+    -> IO (Either ProcessHasExited a)
+withCluster tr severity n action = do
+    ports <- randomUnusedTCPPorts (n + 1)
+    print $ permutations ports
```
probably we want to use our logging infrastructure someday in the future - if not doing this now maybe add TO_DO to not miss it later

KtorZ replied:
What do you mean?

paweljakubas replied:
I mean, we have now `print` - and maybe use our logging here?

Anviking replied:
Also, something more descriptive than
> [(32403,[])]

would be nice

KtorZ replied:
Ah sorry. I was really confused by your comment at first and didn't get why. This line is actually an oversight. I removed it :+1:

paweljakubas commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -163,152 +179,250 @@ parseGenesisData = \case
+    --
+    -- >>> permutations [1,2,3]
+    -- [(1,[2,3]), (2, [1,3]), (3, [1,2])]
+    permutations :: Eq a => [a] -> [(a, [a])]
```
maybe reuse https://hackage.haskell.org/package/base-4.14.0.0/docs/src/Data.OldList.html#permutations

paweljakubas replied:
and then fmap with `f(x:xs)=[x,[xs]]` ?

KtorZ replied:
It won't work because it's does not actually compute _all_ permutations (unlike Data.List.permutations). Using Data.List.permutations on the example above, we would also get [(1, [3,2]), (2, [3,1]), (3, [2,1])]. So, we should also nub + sort the second element which is arguably simpler than the current function :thinking: .. 

Changed for:

```hs
    rotate :: Ord a => [a] -> [(a, [a])]
    rotate = nub . fmap (\(x:xs) -> (x, sort xs)) . permutations
```

paweljakubas reviewed (approved):
looks like sound scaffolding for what is coming next

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -163,152 +179,250 @@ parseGenesisData = \case
+
+        forM_ (init $ permutations ports) $ \(port, peers) -> do
+            withAsync
+                (withStakePool tr severity (port, peers) $ readChan doneGroup)
```
Why not put the n-th `withStakePool` in the n-1-th's action?

Will all nodes stop if on stake pool goes down?

KtorZ replied:
> Why not put the n-th withStakePool in the n-1-th's action?

What?

> Will all nodes stop if on stake pool goes down?

No. If one pools doesn't start for some reason, others will be cancelled. But once they've all started, if one dies, the cluster will keep running.

Anviking replied:
> What?

Something like this:

```haskell
withStakePool args $ do
    withStakePool args $ do
        withStakePool args $ do
            .
              . n times
                .
```

Slower to start, but maybe simpler, and should stop all if one goes down.

Anviking reviewed (approved):
looks neat

KtorZ committed:
rewrite 'permutations' using built-in composition from Data.List

KtorZ committed:
fix hlint suggestions

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -163,152 +179,242 @@ parseGenesisData = \case
+        group <- waitAll
+        if length (catMaybes group) /= n then do
+            cancelAll
+            throwIO $ ProcessHasExited
```
This looks like it is mixing thrown exceptions with `IO (Either e a)`. If the bft node fails, it will return `Left`. If the others fail to start, they will throw.

KtorZ replied:
Yes, I've changed that later actually while continuing working on this, but this isn't visible from this PR _yet_

rvl commented on lib/shelley/test/data/cardano-node-shelley/genesis.yaml:
```diff
@@ -34,10 +34,10 @@ initialFunds: {}
 staking:
 slotsPerKESPeriod: 86400
-slotLength: 8
+slotLength: 0.2
```
Wow, that's a short slot. So if the tests get paused for 100x0.2=20secs (for whatever reason), they will not be able to continue.

KtorZ replied:
I've tweaked according to what the ledger / consensus teams are using for testing as well; it seems fairly okay though. I don't get why the tests wouldn't be able to continue on if they pause for 20s :thinking: ?

rvl replied:
The epoch length is 20s and I _think_ if an entire epoch's worth of slots are missed then it will not be able to elect slot leaders for the next epoch. But it should be ok anyway. Nobody's going to press ctrl-z during integration tests and expect them to pass.

KtorZ replied:
The cluster is made of one BFT leader + 3 pools though, so there's always the BFT leader capable of producing blocks. I am not sure how that influence the slot leader election however :thinking:

KtorZ replied:
> Nobody's going to press ctrl-z during integration tests and expect them to pass.

I guess that's also a fair assumption :)

rvl commented on lib/core/src/Cardano/Wallet/Network/Ports.hs:
```diff
@@ -26,16 +26,20 @@ module Cardano.Wallet.Network.Ports
     -- * Helpers
     , waitForPort
     , unsafePortNumber
+    , findPort
+    , randomUnusedTCPPorts
```
Ideally we wouldn't have test util functions in the core library.
But I think you couldn't put it into the core-integration library due to a package cycle. Is that correct?

KtorZ replied:
Actually no. I needed to share them and `Cardano.Wallet.Network.Ports` felt like a logical place to put them. I see your point about the "non production" aspect of these two. I believe they could fit in core-integration as well :thinking:

KtorZ committed:
use smaller delta for transaction list test.

One second is actually pretty large and we ended up in cases where,
adding one second to the time of the first transaction would make it
later than the second transaction! Causing some tests to fail for a
wrong reason. This delta needs to be at most the slot length, so I've
made it 100ms.
